File: src/Moka.Red.Forms/SignaturePad/MokaSignaturePad.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using Moka.Red.Core.Base;
using Moka.Red.Core.Utilities;

namespace Moka.Red.Forms.SignaturePad;

/// <summary>
///     Drawing canvas for capturing signatures. Uses JS interop for canvas drawing operations.
///     Supports undo, clear, and exports the signature as a base64 data URL. A <see cref="Value" />
///     from the parent is drawn on the canvas, and an empty one clears it.
/// </summary>
public partial class MokaSignaturePad : MokaVisualComponentBase
{
	private const string ModulePath = "./_content/Moka.Red.Forms/SignaturePad/MokaSignaturePad.razor.js";

	private ElementReference _canvasRef;
	private DotNetObjectReference<MokaSignaturePad>? _dotNetRef;
	private bool _drawPending;
	private int _drawRequest;
	private bool _initialized;
	private string? _lastValue;

	// The value from the parent the canvas shows under the user's strokes. Undoing the last of them
	// reports no drawing, but the canvas shows this value again.
	private string? _shownValue;
	private string? _value;

	/// <summary>
	///     The signature as a base64 data URL. Two-way bindable. A value the parent passes is drawn on
	///     the canvas, scaled to fit, and a null or empty one clears it.
	/// </summary>
	[Parameter]
	public string? Value { get; set; }

	/// <summary>Callback when the signature value changes.</summary>
	[Parameter]
	public EventCallback<string?> ValueChanged { get; set; }

	/// <summary>Label text displayed above the pad.</summary>
	[Parameter]
	public string? Label { get; set; }

	/// <summary>Helper text displayed below the pad.</summary>
	[Parameter]
	public string? HelperText { get; set; }

	/// <summary>Width of the signature pad. Default "100%".</summary>
	[Parameter]
	public string Width { get; set; } = "100%";

	/// <summary>Height of the signature pad. Default "200px".</summary>
	[Parameter]
	public string Height { get; set; } = "200px";

	/// <summary>Stroke color for drawing. Default "#000000".</summary>
	[Parameter]
	public string StrokeColor { get; set; } = "#000000";

	/// <summary>Stroke width in pixels. Default 2.</summary>
	[Parameter]
	public int StrokeWidth { get; set; } = 2;

	/// <summary>Background color of the canvas. Default "#ffffff".</summary>
	[Parameter]
	public string BackgroundColor { get; set; } = "#ffffff";

	/// <summary>Whether the pad is read-only.</summary>
	[Parameter]
	public bool ReadOnly { get; set; }

	/// <summary>Whether to show the clear button. Default true.</summary>
	[Parameter]
	public bool ShowClearButton { get; set; } = true;

	/// <summary>Whether to show the undo button. Default true.</summary>
	[Parameter]
	public bool ShowUndoButton { get; set; } = true;

	/// <summary>Placeholder text shown when the pad is empty. Default "Sign here".</summary>
	[Parameter]
	public string Placeholder { get; set; } = "Sign here";

	/// <inheritdoc />
	protected override string RootClass => "moka-signature-pad";

	/// <inheritdoc />
	protected override string CssClass => new CssBuilder(RootClass)
		.AddClass("moka-signature-pad--disabled", Disabled)
		.AddClass("moka-signature-pad--readonly", ReadOnly)
		.AddClass(Class)
		.Build();

	// The root is the outermost element, so it takes the margin. The canvas frame draws the pad's
	// border, so it takes the padding and the radius.

	/// <inheritdoc />
	protected override string? CssStyle => new StyleBuilder()
		.AddStyle("width", Width)
		.AddStyle("margin", ResolvedMargin)
		.AddStyle(Style)
		.Build();

	private string? FrameStyle => new StyleBuilder()
		.AddStyle("padding", ResolvedPadding)
		.AddStyle("border-radius", ResolvedRounding)
		.Build();

	// Height was documented but never applied, so the canvas took the 3:1 shape of its drawing
	// buffer whatever the parameter said.
	private string? CanvasStyle => new StyleBuilder()
		.AddStyle("height", Height)
		.AddStyle("background-color", BackgroundColor)
		.Build();

	/// <summary>Override ShouldRender to always return true for canvas state updates.</summary>
	protected override bool ShouldRender() => true;

	/// <inheritdoc />
	protected override void OnParametersSet()
	{
		base.OnParametersSet();

		// Value only seeds the state when the parent passes a new value. Copying it on every parent
		// render put the "Sign here" placeholder back over an unbound pad's signature.
		if (!string.Equals(_lastValue, Value, StringComparison.Ordinal))
		{
			_lastValue = Value;

			// A bound parent passes back the value the pad just reported. That drawing is already on
			// the canvas, and drawing it again would throw away the strokes undo works through.
			if (!string.Equals(_value, Value, StringComparison.Ordinal))
			{
				// null and "" both mean an empty pad, so one following the other clears nothing.
				if (!string.IsNullOrEmpty(_value) || !string.IsNullOrEmpty(Value))
				{
					_drawPending = true;
				}

				_value = Value;
			}
		}
	}

	/// <inheritdoc />
	protected override async Task OnAfterRenderAsync(bool firstRender)
	{
		// The drawing buffer takes the displayed shape once, before anything is painted on it.
		if (firstRender)
		{
			await SafeModuleInvokeVoidAsync(ModulePath, "fitCanvas", _canvasRef);
		}

		// Drawing is attached the first time the pad is editable: a pad that started read-only or
		// disabled could never be drawn on. Attaching paints the background, so a signature already
		// on the canvas is drawn again after it.
		if (!_initialized && !ReadOnly && !Disabled)
		{
			await InitializeCanvasAsync();
			if (!firstRender && !string.IsNullOrEmpty(_value))
			{
				_drawPending = true;
			}
		}

		if (_drawPending)
		{
			_drawPending = false;
			int request = ++_drawRequest;
			string? value = _value;
			bool shown = await SafeModuleInvokeAsync<bool>(ModulePath, "showSignature", _canvasRef, value,
				BackgroundColor);

			// A newer value may have been sent while the image loaded. Only the latest draw counts.
			if (request == _drawRequest)
			{
				_shownValue = shown ? value : null;
			}
		}
	}

	private async Task InitializeCanvasAsync()
	{
		if (_initialized)
		{
			return;
		}

		_dotNetRef = DotNetObjectReference.Create(this);

		// A dictionary rather than an anonymous type, which trimming can strip in WebAssembly apps.
		await SafeModuleInvokeVoidAsync(ModulePath, "initCanvasDraw", _dotNetRef, _canvasRef,
			new Dictionary<string, object?>
			{
				["strokeColor"] = StrokeColor,
				["strokeWidth"] = StrokeWidth,
				["backgroundColor"] = BackgroundColor
			});
		_initialized = true;
	}

	/// <summary>Called from JavaScript when the signature changes.</summary>
	[JSInvokable]
	public async Task OnSignatureChanged(string? signatureData)
	{
		_value = signatureData ?? _shownValue;

		// A call from JS renders nothing by itself, and without a bound parent nothing else would
		// hide the placeholder.
		StateHasChanged();

		if (ValueChanged.HasDelegate)
		{
			await ValueChanged.InvokeAsync(_value);
		}
	}

	private async Task HandleClear()
	{
		if (!_initialized)
		{
			return;
		}

		// A value still loading is not drawn over the cleared pad, in JS or here.
		_drawRequest++;
		await SafeModuleInvokeVoidAsync(ModulePath, "clearSignature", _canvasRef, BackgroundColor);
		_value = null;
		_shownValue = null;
		await ValueChanged.InvokeAsync(null);
	}

	private async Task HandleUndo()
	{
		if (!_initialized)
		{
			return;
		}

		await SafeModuleInvokeVoidAsync(ModulePath, "undoSignature", _canvasRef, StrokeColor, StrokeWidth,
			BackgroundColor);
	}

	/// <inheritdoc />
	protected override async ValueTask DisposeAsyncCore()
	{
		if (_initialized)
		{
			await SafeModuleInvokeVoidAsync(ModulePath, "removeCanvasDraw", _canvasRef);
		}

		_dotNetRef?.Dispose();
		await base.DisposeAsyncCore();
	}
}

File: src/Moka.Red.Forms/SignaturePad/MokaSignaturePad.razor.js
/**
 * MokaSignaturePad - shows a signature the parent passes as Value.
 *
 * Drawing with the pointer lives in Core's moka-drag.js and is re-exported here, so the component
 * only ever imports this one module (MokaComponentBase caches a single module per component).
 * The relative import resolves against this module's own URL, so it keeps working when the host
 * app is served from a sub-path.
 */

import { initCanvasDraw, clearCanvas, undoCanvas, removeCanvasDraw } from '../../Moka.Red.Core/moka-drag.js';

export { initCanvasDraw, removeCanvasDraw };

/**
 * Gives the drawing buffer the shape the canvas is displayed at: the width stays (600, so a saved
 * signature keeps its size) and the height follows the displayed aspect ratio. Without it a Height
 * other than a third of the width would stretch every stroke.
 * @param {HTMLCanvasElement} canvas
 */
export function fitCanvas(canvas) {
	if (!canvas) return;

	const rect = canvas.getBoundingClientRect();
	if (rect.width < 1 || rect.height < 1) return;

	const height = Math.max(1, Math.round(canvas.width * rect.height / rect.width));
	if (canvas.height !== height) canvas.height = height;
}

// canvas -> the background plus the signature from Value, painted at the canvas size.
const layers = new WeakMap();

// canvas -> the latest showSignature call. An image that loads after a newer call must not paint.
const requests = new WeakMap();

/**
 * Replaces the drawing with a signature: the data URL is loaded into an image and drawn scaled to
 * fit the canvas, centred, over the background. An empty value clears the canvas. The strokes drawn
 * so far are dropped either way. Works on a read-only canvas too, where drawing was never attached.
 * @param {HTMLCanvasElement} canvas
 * @param {?string} dataUrl - The signature, usually a PNG data URL the pad produced before.
 * @param {string} backgroundColor
 * @returns {Promise<boolean>} Whether a signature is shown. False for an empty value or an image
 *   that did not load, and for a call a newer one replaced.
 */
export function showSignature(canvas, dataUrl, backgroundColor) {
	if (!canvas) return Promise.resolve(false);

	const request = {};
	requests.set(canvas, request);

	if (!dataUrl) {
		clearSignature(canvas, backgroundColor);
		return Promise.resolve(false);
	}

	return new Promise(resolve => {
		const image = new Image();
		image.onload = () => {
			if (requests.get(canvas) !== request) {
				resolve(false);
				return;
			}

			const layer = paintLayer(canvas, image, backgroundColor);
			clearSignature(canvas, backgroundColor);
			canvas.getContext('2d').drawImage(layer, 0, 0);
			layers.set(canvas, layer);
			resolve(true);
		};
		image.onerror = () => {
			// A value that is not an image shows as an empty pad rather than as the old drawing.
			if (requests.get(canvas) === request) clearSignature(canvas, backgroundColor);
			resolve(false);
		};

		// An image from another origin without CORS headers would taint the canvas, and the next
		// stroke's toDataURL would throw. This way it fails to load instead. Data URLs are unaffected.
		image.crossOrigin = 'anonymous';
		image.src = dataUrl;
	});
}

/**
 * Clears the canvas to the background and forgets both the strokes and a signature from Value.
 * An image still loading for an earlier showSignature call is not drawn when it arrives.
 * @param {HTMLCanvasElement} canvas
 * @param {string} backgroundColor
 */
export function clearSignature(canvas, backgroundColor) {
	if (!canvas) return;

	requests.delete(canvas);
	layers.delete(canvas);

	// Forgets the strokes too, when drawing is attached. A read-only canvas has none.
	clearCanvas(canvas, backgroundColor);
	fill(canvas, backgroundColor);
}

/**
 * Removes the last stroke. moka-drag.js repaints the background and then the strokes that are
 * left, which would wipe a signature from Value, so that signature is handed over as the
 * background: a pattern of the painted layer fills the canvas exactly.
 * @param {HTMLCanvasElement} canvas
 * @param {string} strokeColor
 * @param {number} strokeWidth
 * @param {string} backgroundColor
 */
export function undoSignature(canvas, strokeColor, strokeWidth, backgroundColor) {
	if (!canvas) return;

	const layer = layers.get(canvas);
	const background = layer ? canvas.getContext('2d').createPattern(layer, 'no-repeat') : backgroundColor;
	undoCanvas(canvas, strokeColor, strokeWidth, background);
}

function paintLayer(canvas, image, backgroundColor) {
	const layer = document.createElement('canvas');
	layer.width = canvas.width;
	layer.height = canvas.height;
	fill(layer, backgroundColor);

	// Fit inside the canvas and keep the aspect ratio, so a signature from a pad of another
	// shape is not stretched. An SVG without a size of its own reports 0 and fills the canvas.
	const imageWidth = image.naturalWidth || layer.width;
	const imageHeight = image.naturalHeight || layer.height;
	const scale = Math.min(layer.width / imageWidth, layer.height / imageHeight);
	const width = imageWidth * scale;
	const height = imageHeight * scale;
	layer.getContext('2d').drawImage(image, (layer.width - width) / 2, (layer.height - height) / 2, width, height);
	return layer;
}

function fill(canvas, backgroundColor) {
	const ctx = canvas.getContext('2d');
	ctx.save();
	ctx.fillStyle = backgroundColor || '#ffffff';
	ctx.fillRect(0, 0, canvas.width, canvas.height);
	ctx.restore();
}
